### ISSUE TITLE
persist: crash when reading parquet file

### ISSUE THREAD

**Issue by danhhz** [MaterializeInc/materialize#10126]
### What version of Materialize are you using?

v0.16.1-dev (bf982adab)

### How did you install Materialize?

Built from source

### What was the issue?

Crash when testing #10096, which is a fix for #10071. Ran #10096 with #10072 cherry-picked on top and the number of records set to 100_000_000. It loaded fine but on restart, persist crashed while starting up

### Relevant log output

```source
2022-01-19T16:11:31.342104Z  INFO materialized: booting server
materialized v0.16.1-dev (bf982adab)
OpenSSL 1.1.1l  24 Aug 2021
librdkafka v1.8.2
invoked as: MZ_SOFT_ASSERTIONS=1 MZ_METRICS_SCRAPING_INTERVAL=1s MZ_DEV=1 materialized '--log-file=stderr' '--data-directory=/share/mzdata' --listen-addr 0.0.0.0:6875 --disable-telemetry --experimental --timestamp-frequency 100ms --introspection-frequency 100ms --retain-prometheus-metrics 1s --persistent-user-tables --persistent-kafka-upsert-source
os: Ubuntu 20.4.0 [64-bit]
cpus: 16 logical, 8 physical, 8 useful
cpu0: AMD EPYC 7571 2531MHz
memory: 133903691KB total, 1410889KB used, 9007199254740988KiB limit
swap: 0KB total, 0KB used, 0KiB limit
dataflow workers: 4
2022-01-19T16:11:31.461788Z  INFO coord::timestamp: Starting Timestamping Thread. Frequency: 10 ms.
materialized v0.16.1-dev (bf982adab) listening on 0.0.0.0:6875...
=======================================================================
Thank you for trying Materialize!

We are interested in any and all feedback you have, which may be able
to improve both our software and your queries! Please reach out at:

    Web: https://materialize.com
    GitHub issues: https://github.com/MaterializeInc/materialize/issues
    Email: support@materialize.com
    Twitter: @MaterializeInc
=======================================================================

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
                                WARNING!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Starting Materialize in experimental mode means:

- This node's catalog of views and sources are unstable.

If you use any version of Materialize besides this one, you might
not be able to start the Materialize node. To fix this, you'll have
to remove all of Materialize's data (e.g. rm -rf mzdata) and start
the node anew.

- You must always start this node in experimental mode; it can no
longer be started in non-experimental/regular mode.

For more details, see https://materialize.com/docs/cli#experimental-mode
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

2022-01-19T16:12:45.180578Z  INFO coord::catalog: create source materialize.public.s1 (u1)
2022-01-19T16:12:45.189545Z  INFO coord::timestamp: Timestamping Source u1 with Real Time Consistency.
2022-01-19T16:12:45.193396Z  INFO coord::timestamp: Discovered 8 new (8 total) kafka partitions for topic testdrive-kafka-recovery-big-1642608690 (source u1)
2022-01-19T16:12:45.198122Z  INFO coord::catalog: create view materialize.public.s1_is_complete (u2)
2022-01-19T16:12:45.198204Z  INFO coord::catalog: create index materialize.public.s1_is_complete_primary_idx (u3)
2022-01-19T16:12:45.214724Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [0] (source u1/29) on worker 2
2022-01-19T16:12:45.215000Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [2] (source u1/29) on worker 0
2022-01-19T16:12:45.215086Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [5] (source u1/29) on worker 3
2022-01-19T16:12:45.215388Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [6] (source u1/29) on worker 0
2022-01-19T16:12:45.215460Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [3] (source u1/29) on worker 1
2022-01-19T16:12:45.215581Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [4] (source u1/29) on worker 2
2022-01-19T16:12:45.215612Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [1] (source u1/29) on worker 3
2022-01-19T16:12:45.215989Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [7] (source u1/29) on worker 1
2022-01-19T16:22:14.066349Z  INFO materialized: booting server
materialized v0.16.1-dev (bf982adab)
OpenSSL 1.1.1l  24 Aug 2021
librdkafka v1.8.2
invoked as: MZ_SOFT_ASSERTIONS=1 MZ_METRICS_SCRAPING_INTERVAL=1s MZ_DEV=1 materialized '--log-file=stderr' '--data-directory=/share/mzdata' --listen-addr 0.0.0.0:6875 --disable-telemetry --experimental --timestamp-frequency 100ms --introspection-frequency 100ms --retain-prometheus-metrics 1s --persistent-user-tables --persistent-kafka-upsert-source
os: Ubuntu 20.4.0 [64-bit]
cpus: 16 logical, 8 physical, 8 useful
cpu0: AMD EPYC 7571 2531MHz
memory: 133903691KB total, 2108684KB used, 9007199254740988KiB limit
swap: 0KB total, 0KB used, 0KiB limit
dataflow workers: 4
2022-01-19T16:22:14.114813Z  INFO coord::catalog: create source materialize.public.s1 (u1)
2022-01-19T16:22:14.115579Z  INFO coord::catalog: create view materialize.public.s1_is_complete (u2)
2022-01-19T16:22:14.115660Z  INFO coord::catalog: create index materialize.public.s1_is_complete_primary_idx (u3)
2022-01-19T16:22:14.122101Z  INFO coord::catalog: create source materialize.public.s1 (u1)
2022-01-19T16:22:14.122870Z  INFO coord::catalog: create view materialize.public.s1_is_complete (u2)
2022-01-19T16:22:14.122951Z  INFO coord::catalog: create index materialize.public.s1_is_complete_primary_idx (u3)
2022-01-19T16:22:14.124613Z  INFO coord::timestamp: Starting Timestamping Thread. Frequency: 10 ms.
2022-01-19T16:22:14.135104Z  INFO coord::timestamp: Timestamping Source u1 with Real Time Consistency.
=======================================================================
Thank you for trying Materialize!

We are interested in any and all feedback you have, which may be able
to improve both our software and your queries! Please reach out at:

    Web: https://materialize.com
    GitHub issues: https://github.com/MaterializeInc/materialize/issues
    Email: support@materialize.com
    Twitter: @MaterializeInc
=======================================================================

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
                                WARNING!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Starting Materialize in experimental mode means:

- This node's catalog of views and sources are unstable.

If you use any version of Materialize besides this one, you might
not be able to start the Materialize node. To fix this, you'll have
to remove all of Materialize's data (e.g. rm -rf mzdata) and start
the node anew.

- You must always start this node in experimental mode; it can no
longer be started in non-experimental/regular mode.

For more details, see https://materialize.com/docs/cli#experimental-mode
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

materialized v0.16.1-dev (bf982adab) listening on 0.0.0.0:6875...
2022-01-19T16:22:14.140980Z  INFO coord::timestamp: Discovered 8 new (8 total) kafka partitions for topic testdrive-kafka-recovery-big-1642608690 (source u1)
2022-01-19T16:26:26.063575Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [4] (source u1/29) on worker 2
2022-01-19T16:26:26.064181Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [0] (source u1/29) on worker 2
2022-01-19T16:26:26.075201Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [3] (source u1/29) on worker 1
2022-01-19T16:26:26.075749Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-kafka-recovery-big-1642608690 [7] (source u1/29) on worker 1
2022-01-19T16:26:36.480214Z ERROR panic: column start and length should not be negative
thread: <unnamed>
location: /cargo/registry/src/github.com-1ecc6299db9ec823/parquet2-0.8.1/src/metadata/column_chunk_metadata.rs:121:9
version: 0.16.1-dev (bf982adab44fb0bb74f74a9901e607cb3e20dca2)
backtrace:
   0: materialized::handle_panic
             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0c57d3390206bc57d-1/materialize/tests/src/materialized/src/bin/materialized/main.rs:838:21
   1: core::ops::function::Fn::call
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:628:17
   3: std::panicking::begin_panic::{{closure}}
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:544:9
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:139:18
   5: std::panicking::begin_panic
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:543:12
   6: parquet2::metadata::column_chunk_metadata::ColumnChunkMetaData::byte_range
             at cargo/registry/src/github.com-1ecc6299db9ec823/parquet2-0.8.1/src/metadata/column_chunk_metadata.rs:121:9
   7: parquet2::read::get_page_iterator
             at cargo/registry/src/github.com-1ecc6299db9ec823/parquet2-0.8.1/src/read/mod.rs:54:26
      <parquet2::read::ColumnIterator<R> as parquet2::read::MutStreamingIterator>::advance
             at cargo/registry/src/github.com-1ecc6299db9ec823/parquet2-0.8.1/src/read/mod.rs:175:20
   8: arrow2::io::parquet::read::column_iter_to_array
             at cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.8.1/src/io/parquet/read/mod.rs:385:15
   9: <arrow2::io::parquet::read::record_batch::RecordReader<R> as core::iter::traits::iterator::Iterator>::next::{{closure}}
             at cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.8.1/src/io/parquet/read/record_batch.rs:153:39
      <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::try_fold::enumerate::{{closure}}
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/iter/adapters/enumerate.rs:85:27
      core::iter::traits::iterator::Iterator::try_fold
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/iter/traits/iterator.rs:1995:21
      <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::try_fold
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/iter/adapters/enumerate.rs:91:9
  10: <arrow2::io::parquet::read::record_batch::RecordReader<R> as core::iter::traits::iterator::Iterator>::next
             at cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.8.1/src/io/parquet/read/record_batch.rs:140:17
  11: persist::indexed::columnar::parquet::decode_parquet_file_kvtd
             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0c57d3390206bc57d-1/materialize/tests/src/persist/src/indexed/columnar/parquet.rs:169:18
  12: persist::indexed::columnar::parquet::decode_trace_parquet
             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0c57d3390206bc57d-1/materialize/tests/src/persist/src/indexed/columnar/parquet.rs:98:42
  13: <persist::indexed::encoding::BlobTraceBatch as persist_types::Codec>::decode
             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0c57d3390206bc57d-1/materialize/tests/src/persist/src/indexed/encoding.rs:630:9
  14: persist::indexed::cache::BlobCache<B>::fetch_trace_batch_sync
             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0c57d3390206bc57d-1/materialize/tests/src/persist/src/indexed/cache.rs:207:37
  15: persist::indexed::cache::BlobCache<B>::get_trace_batch_async::{{closure}}
             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0c57d3390206bc57d-1/materialize/tests/src/persist/src/indexed/cache.rs:246:23
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:123:18
  16: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/mod.rs:483:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:403:40
      std::panicking::try
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:367:19
      std::panic::catch_unwind
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panic.rs:133:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/mod.rs:482:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
  17: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/boxed.rs:1691:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/boxed.rs:1691:9
      std::sys::unix::thread::Thread::new::thread_start
             at rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys/unix/thread.rs:106:17
  18: start_thread
  19: clone

materialized encountered an internal error and crashed.

We rely on bug reports to diagnose and fix these errors. Please
copy and paste the above details and file a report at:

    https://materialize.com/s/bug
```
```

**Comment by danhhz** [comment]
Tried to repro the issue with various sizes of the end_to_end benchmark, but was unsuccessful. Downloaded the data, fired up a local build of mz with some extra debugging info and didn't see a panic. It doesn't look like it's loading all the batches I downloaded, now investigating that...

**Comment by danhhz** [comment]
Hmm now possibly seeing it start decoding a batch but never come back? That's a different issue than a panic, but maybe it's because I'm not running with --release but the original issue was. Will try --release now

**Comment by danhhz** [comment]
Ah! That is indeed because it's not --release. It turns out that BlobTraceBatch::validate is quite expensive without --release

**Comment by danhhz** [comment]
Yay got a repro! As I suspected, it's in a large batch, so presumably one of the new ones that 10096 introduced that have multiple parquet batches in it

**Comment by danhhz** [comment]
> column start 972445496 and length -1985409397 should not be negative

seems like an overflow or something

**Comment by danhhz** [comment]
Looks like this is wrong in the parquet metadata itself

```
                ColumnChunkMetaData {
                    column_chunk: ColumnChunk {
                        file_path: None,
                        file_offset: -1012963901,
                        meta_data: Some(
                            ColumnMetaData {
                                type_: Type(
                                    6,
                                ),
                                encodings: [
                                    Encoding(
                                        3,
                                    ),
                                    Encoding(
                                        0,
                                    ),
                                ],
                                path_in_schema: [
                                    "v",
                                ],
                                codec: CompressionCodec(
                                    0,
                                ),
                                num_values: 40518559,
                                total_uncompressed_size: -1985409397,
                                total_compressed_size: -1985409397,
                                key_value_metadata: None,
                                data_page_offset: 972445496,
                                index_page_offset: None,
                                dictionary_page_offset: None,
                                statistics: None,
                                encoding_stats: None,
                                bloom_filter_offset: None,
                            },
                        ),
                        offset_index_offset: None,
                        offset_index_length: None,
                        column_index_offset: None,
                        column_index_length: None,
                        crypto_metadata: None,
                        encrypted_column_metadata: None,
                    },
```

**Comment by danhhz** [comment]
Okay, I think I've figured this out. First, a bit of (probably unnecessary) background. A parquet file has a schema with a set of declared columns. The data part is broken up first into a series of _row groups_. Each of these row groups have a schema that must match the file schema (dunno why). Each row group contains one _column chunk_ per column in the file/groups's schema. Each column chunk is made up of 1 or more _pages_. A page is the unit of compression.

I'm reasonably certain the bug is these two lines: https://github.com/jorgecarleitao/parquet2/blob/v0.8.1/src/write/page.rs#L107-L108

```
        uncompressed_page_size: compressed_page.uncompressed_size() as i32,
        compressed_page_size: compressed_page.compressed_size() as i32,
```

We (currently, see below) get 1 parquet page for each arrow2 `Array` (1:1 with our (K, V, T, D) columns) that we encode. There's nothing I can see that keeps these `uncompressed_size` or `compressed_size` methods from overflowing an i32, which ends up causing these two fields to go negative on the parquet PageHeader metadata. This ends up being used in parquet2 to compute the other sizes stored at higher levels of metadata, which is what was originally triggering the panic.

parquet2 currently maps an Array to a page 1:1 (), but this doesn't have to be the case. It's free to break up an Array into as many pages as necessary to keep each under the relevant size constraints and probably should do this behind the scenes. This bit happens right now in arrow2 via the `RowGroupIterator`, ultimately hitting this line which always maps a data Array into one page: https://github.com/jorgecarleitao/arrow2/blob/v0.8.1/src/io/parquet/write/mod.rs#L129-L130

```
        _ => array_to_page(array, descriptor, options, encoding)
            .map(|page| DynIter::new(std::iter::once(Ok(page)))),
```

I'd imagine that array_to_page should break up the array into as many pages as necessary.

I should be able to construct an easy repro and file a bug, which I'll do. In the meantime, to work around this, I think it's sufficient to make sure we never create ColumnarRecords such that any Array takes up more than i32::MAX bytes when encoded. For K and V, we already make sure the data part of the columnar byte arrays is <= i32::MAX, but I think the page ends up also encoding the offsets too, which could push us over.

**Comment by danhhz** [comment]
Closed by #10096
